### DISCUSSION TITLE
Dependency-propagating Schedulers

### DISCUSSION THREAD

**Discussion by kabiroberai** [pointfreeco/swift-dependencies#85] [Ideas]
**Update**: There's now a PR for this: see #86.

Scheduler APIs currently fail to propagate dependencies across the `schedule` boundary:
```swift
mainQueue.schedule {
  // dependencies from outside this closure aren't brought inside
}

values
  .receive(on: mainQueue)
  .sink { ... } // same issue
```
If swift-dependencies didn't claim to support schedulers at all this would _maybe_ be acceptable, but there's even a `mainQueue` dependency so it's extra strange that using it doesn't propagate deps. Plus, it'd be nice if we eased the transition from Combine to Concurrency by somewhat supporting Dependencies in the latter.

One approach to solve this is to add a `dependencies` modifier to `Scheduler`, similar to `animation`/`transaction`:
```swift
extension Scheduler {
  /// Propagates dependencies across the scheduler boundary.
  ///
  /// - Parameter update: Enables transforming the propagated dependencies. No-ops by default.
  func dependencies(
    _ update: @escaping (inout DependencyValues) -> Void = { _ in }
  ) -> AnySchedulerOf<Self>
}
```
Under the hood this would wrap the `schedule` calls in `withEscapedDependencies` to propagate the continuation.

We could also make all built-in CombineSchedulers (e.g. `UIScheduler`) forward dependencies by default, or at least do so for the `mainQueue` dependency. Would be happy to put up a PR to 1) add `Scheduler.dependencies` and 2) forward dependencies on `mainQueue` — curious to hear other people's thoughts.

**Comment by stephencelis** [comment]
@kabiroberai swift-dependencies depends on combine-schedulers and swift-clocks, and not the other way around, so I'm not sure it's really possible for us to propagate dependencies _automatically_ without combining these libraries.

Your `Scheduler.dependencies` operation is definitely interesting, though! We'll have a chat about it soon :smile:

**Reply by kabiroberai** [reply]
Yeah agreed that the UIScheduler feature would be a bit tricky and might require some sort of dependency inversion. On the other hand, since we declare the `mainQueue` dependency inside this package I think we _can_ make that propagate dependencies, and if you have Dependencies as a dependency (hah) you shouldn't be using `UIScheduler.shared` directly anyway.

**Reply by mbrandonw** [reply]
Hey @kabiroberai, we think this looks pretty good, so we will probably move forward with it, but we do need a bit more time to review it. We'll be in touch!
